docs/tutorials/quantum_reinforcement_learning.ipynb

```diff
@@ -0,0 +1,4085 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this tutorial, we describe how to implement a PQC with data re-uploading in TFQ, and use it as:\n",
+    "1. an RL policy trained with a policy-gradient method,\n",
+    "2. a Q-function approximator trained with deep Q-learning,\n",
+    "\n",
+    "each solving [CartPole-v1](http://gym.openai.com/envs/CartPole-v1/), a benchmarking task from OpenAI Gym. Note that, as showcased in <a href=\"https://arxiv.org/abs/2103.05577\" class=\"external\">[1]</a> and <a href=\"https://arxiv.org/abs/2103.15084\" class=\"external\">[2]</a>, these implementations can also be used to solve other task-environment from OpenAI Gym, such as [FrozenLake-v0](http://gym.openai.com/envs/FrozenLake-v0/), [MountainCar-v0](http://gym.openai.com/envs/MountainCar-v0/) or [Acrobot-v1](http://gym.openai.com/envs/Acrobot-v1/)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Features of this implementation:\n",
+    "- it shows how to use a `tfq.layers.ControlledPQC` to implement a PQC with data re-uploading, appearing in many applications of QML. This implementation also naturally allows using trainable scaling parameters at the input of the PQC, to increase its expressivity,\n",
+    "- it shows how to implement observables with trainable weights at the output of a PQC, to allow a flexible range of output values,\n",
+    "- it shows how a `tf.keras.Model` can be trained with non-trivial ML loss functions, i.e., that are not compatible with `model.compile` and `model.fit`, using a `tf.GradientTape`."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setup"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**<!>** To avoid automatic multiprocessing of TensorFlow, which can render the execution of this notebook slower, we recommend limiting the number of CPUs accessible to Jupyter to 1, e.g., by running the following command on Linux `taskset --cpu-list 0 jupyter notebook` when starting Jupyter."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Install TensorFlow:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install -q tensorflow==2.3.1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Install TensorFlow Quantum:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install -q tensorflow-quantum"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Install Gym:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install gym"
```

Comment:
Should probably pick a particular version here, in case of breaking updates/changes down the line.